doc/user: Add metrics and service-monitor docs

LiliC opened the pull request:
<!--

Before making a PR, please read our contributing guidelines https://github.com/operator-framework/operator-sdk/blob/master/CONTRIBUTING.MD

Note: Make sure your branch is rebased to the latest upstream master.

-->

**Description of the change:**

Basic documentation for monitoring.

LiliC commented:
As agreed offline: we want to also a `doc.go` file to the `pkg/metrics`.

LiliC commented:
@shawn-hurley You mentioned you wanted me to add a `doc.go` file to the `pkg/metrics/`, tried doing that but think I would just duplicate what it says here. Ideas welcome!

LiliC commented:
Blocked until https://github.com/operator-framework/operator-sdk/pull/1037 is merged, but ready for review.

LiliC commented:
Happy with all the suggestions here, docs are not my strong point. :)

joelanford commented on doc/user/metrics/README.md:
```diff
@@ -0,0 +1,44 @@
+
+The `func ExposeMetricsPort(ctx context.Context, port int32) (*v1.Service, error)` is a helper function that exposes general metrics about the running program. These metrics are inherited from controller-runtime. The metrics are by default served on `:8383/metrics`. A [Service][service] object is created with the metrics port exposed, which can be then accessed by Prometheus. The Service object i ...
+
+To modify the port the metrics are exposed on, change the variable `metricsHost string = ":8383"` variable in the `cmd/manager/main.go` file of the generated operator.
```
I would suggest moving the sentence in the above paragraph to the first sentence of this paragraph.

```suggestion
By default, the metrics are served on `0.0.0.0:8383/metrics`. To modify the port the metrics are exposed on, change the `var metricsPort int32 = 8383` variable in the `cmd/manager/main.go` file of the generated operator.
```

Should we mention anything about changing (or not changing) the `metricsHost` variable?

LiliC replied:
> I would suggest moving the sentence in the above paragraph to the first sentence of this paragraph.

Sorry, not sure I understand that. :)

joelanford commented on doc/user/metrics/README.md:
```diff
@@ -0,0 +1,44 @@
+
+## Metrics in operator-sdk
+
+The `func ExposeMetricsPort(ctx context.Context, port int32) (*v1.Service, error)` is a helper function that exposes general metrics about the running program. These metrics are inherited from controller-runtime. The metrics are by default served on `:8383/metrics`. A [Service][service] object is created with the metrics port exposed, which can be then accessed by Prometheus. The Service object i ...
```
```suggestion
The `func ExposeMetricsPort(ctx context.Context, port int32) (*v1.Service, error)` function exposes general metrics about the running program. These metrics are inherited from controller-runtime. This helper function creates a [Service][service] object with the metrics port exposed, which can then be accessed by Prometheus. The Service object is [garbage collected][gc] when the leader pod's root owner is deleted.
```

Do you think we need an example for the "leader pod's root owner", or whatever wording we end up with on that?

LiliC replied:
An example sure, what were you thinking of?

joelanford commented on doc/user/metrics/service_monitor.md:
```diff
@@ -0,0 +1,28 @@
+[prometheus-operator][prom-operator] an operator that creates, configures and manages Prometheus clusters atop Kubernetes.
+
+`ServiceMonitor` is a [CR][cr] of the prometheus-operator, which discovers the `Endpoints` in `Service` objects and configures Prometheus to monitor those `Pod`s. The `GenerateServiceMonitor` takes a `Service` object and generates a `ServiceMonitor` resource based on it.
+To add `Service` target discovery of your created monitoring `Service`s you can use the `metrics.CreateServiceMonitor()` helper function, which accepts the newly created `Service`.
```
We should probably add some context about when it might be useful to use a service monitor. Is this something users are supposed to create to monitor their operator, resources created by their operator, or both?

LiliC replied:
As ServiceMonitor is strictly a prom-operator concept, I linked to their architecture docs on this. I think those who are interested in creating the object probably already know why they want it? The general use case is most likely when you have a lot of Services and use the operator.

joelanford commented on doc/user/metrics/service_monitor.md:
````diff
@@ -0,0 +1,28 @@
+    }
+```
+
+*Note:* Create one `ServiceMonitor` per application and per `Namespace`.
````
Maybe expand on this a little bit.

LiliC replied:
Moved this as a comment in code example so it hopefully makes more sense.

LiliC commented:
Renamed the file as well to `service-monitor.md` We have mixed file names but managed to find a style guide that mentions file naming conventions for markdown, and according to that hyphens are the way to go -> http://www.cirosantilli.com/markdown-style-guide/#file-name

LiliC commented:
@joelanford Made some adjustments, can you please have another look.

> Also, should we link to these from the user guide?

Yes, was thinking the same but wasn't sure where it should be placed, if you have any suggestions?

joelanford commented on doc/user/metrics/service-monitor.md:
```diff
@@ -0,0 +1,38 @@
+    serviceMonitors, err := metrics.CreateServiceMonitors(restConfig, ns, services)
+    if err != nil {
+        // handle error here
+    }
```
Would the above also typically be called in `func main()`? If, so I'd suggest also wrapping this in `func main()` opening and closing.

joelanford commented:
> Yes, was thinking the same but wasn't sure where it should be placed, if you have any suggestions?

@LiliC I'd say under the "Advanced Topics" section would be a good spot. WDYT?

https://github.com/operator-framework/operator-sdk/blob/master/doc/user-guide.md#advanced-topics

LiliC commented on doc/user/metrics/README.md:
```diff
@@ -0,0 +1,44 @@
+	if err != nil {
+        // handle error
+		log.Info(err.Error())
+	}
```
I did the single dots with no comments in case someone copies this so they know it's meant to be added, comments would be ignored, hope that makes sense.

LiliC commented:
> I'd say under the "Advanced Topics" section would be a good spot. WDYT?

Done, also addressed the rest of the comments, PTALA, thanks!

joelanford commented:
LGTM :+1:

LiliC commented:
Pinging again for another review :) @AlexNPavel @hasbro17 @theishshah @estroz

AlexNPavel commented on doc/user/metrics/service-monitor.md:
```diff
@@ -0,0 +1,45 @@
+
+`ServiceMonitor` is a [CR][cr] of the prometheus-operator, which discovers the `Endpoints` in `Service` objects and configures Prometheus to monitor those `Pod`s. See the prometheus-operator [documention][sm] to learn more about `ServiceMonitor`s.
+
+The `GenerateServiceMonitor` takes a `Service` object and generates a `ServiceMonitor` resource based on it. To add `Service` target discovery of your created monitoring `Service`s you can use the `metrics.CreateServiceMonitor()` helper function, which accepts the newly created `Service`.
```
```suggestion
The `GenerateServiceMonitor` function takes a `Service` object and generates a `ServiceMonitor` resource based on it. To add `Service` target discovery of your created monitoring `Service`s you can use the `metrics.CreateServiceMonitor()` helper function, which accepts the newly created `Service`.
```

AlexNPavel commented on doc/user/metrics/service-monitor.md:
```diff
@@ -0,0 +1,45 @@
+
+### Prerequisites:
+
+- [prometheus-operator][prom-quickstart] needs to be deployed already in the cluster.
```
```suggestion
- [prometheus-operator][prom-quickstart] needs to be deployed in the cluster.
```

AlexNPavel commented on doc/user/metrics/service-monitor.md:
```diff
@@ -0,0 +1,45 @@
+## Using the ServiceMonitor prometheus-operator CRD
+
+[prometheus-operator][prom-operator] an operator that creates, configures and manages Prometheus clusters atop Kubernetes.
```
```suggestion
[prometheus-operator][prom-operator] is an operator that creates, configures, and manages Prometheus clusters atop Kubernetes.
```

hasbro17 commented on doc/user-guide.md:
```diff
@@ -407,6 +407,10 @@ To implement complex deletion logic, you can add a finalizer to your Custom Reso
 
+### Metrics
+
+To learn about how metrics works in operator-sdk read the [metrics section][metrics_doc] of the user documentation.
```
Nit: metrics is plural so we drop the s for the plural verb work.
```suggestion
To learn about how metrics work in the Operator SDK read the [metrics section][metrics_doc] of the user documentation.
```
Also unless we're talking about the cmd line tool we refer to it as the Operator SDK in general.

hasbro17 commented on doc/user/metrics/service-monitor.md:
```diff
@@ -0,0 +1,45 @@
+
+[prometheus-operator][prom-operator] an operator that creates, configures and manages Prometheus clusters atop Kubernetes.
+
+`ServiceMonitor` is a [CR][cr] of the prometheus-operator, which discovers the `Endpoints` in `Service` objects and configures Prometheus to monitor those `Pod`s. See the prometheus-operator [documention][sm] to learn more about `ServiceMonitor`s.
```
Typo. Also change the link tag from sm to service-monitor.
Mostly for our sake but at a glance it's more clear what we're linking to.
```suggestion
`ServiceMonitor` is a [CR][cr] of the prometheus-operator, which discovers the `Endpoints` in `Service` objects and configures Prometheus to monitor those `Pod`s. See the prometheus-operator [documentation][service-monitor] to learn more about `ServiceMonitor`s.
```

LiliC commented:
Done, PTALA @hasbro17

hasbro17 commented on doc/user/metrics/README.md:
````diff
@@ -0,0 +1,51 @@
+```go
+    import(
+        "github.com/operator-framework/operator-sdk/pkg/metrics"
+	    "sigs.k8s.io/controller-runtime/pkg/manager"
````
Fix indentation to align with the other import:
```suggestion
        "sigs.k8s.io/controller-runtime/pkg/manager"
```

hasbro17 commented on doc/user/metrics/README.md:
````diff
@@ -0,0 +1,51 @@
+    }
+```
+
+Note: The above example is already present in `cmd/manager/main.go` in all the operators generated with operator-sdk.
````
Nit: Just to highlight the note.
```suggestion
*Note:* The above example is already present in `cmd/manager/main.go` in all the operators generated with operator-sdk.
```

hasbro17 commented on doc/user/metrics/service-monitor.md:
```diff
@@ -0,0 +1,45 @@
+
+[prometheus-operator][prom-operator] is an operator that creates, configures, and manages Prometheus clusters atop Kubernetes.
+
+`ServiceMonitor` is a [CR][cr] of the prometheus-operator, which discovers the `Endpoints` in `Service` objects and configures Prometheus to monitor those `Pod`s. See the prometheus-operator [documention][service-monitor] to learn more about `ServiceMonitor`s.
```
CR==>CustomResource.
I think it's better to use the full name if we haven't defined an acronym before using it in a given doc.
```suggestion
`ServiceMonitor` is a CustomResource of the prometheus-operator, which discovers the `Endpoints` in `Service` objects and configures Prometheus to monitor those `Pod`s. See the prometheus-operator [documention][service-monitor] to learn more about `ServiceMonitor`s.
```
Also just a suggestion but we probably can do without linking to the CustomResource docs here. At this point users should be familiar with CR and CRDs.
Linking to the ServiceMonitor doc is all we need which we've already done.

hasbro17 replied:
Maybe this is subjective but "`Pod`s" looks a bit odd to me.
Can we change it to either "`Pods`" or just "Pods"(without the highlighting)?
Same for "`Service`s" and "`ServiceMonitor`s"

hasbro17 commented on doc/user/metrics/service-monitor.md:
````diff
@@ -0,0 +1,45 @@
+
+```go
+    import(
+        v1 "k8s.io/api/core/v1"
````
Is the alias `v1` needed if the package name is already `v1`? 
```suggestion
        "k8s.io/api/core/v1"
```

hasbro17 commented on doc/user/metrics/service-monitor.md:
```diff
@@ -0,0 +1,45 @@
+## Using the ServiceMonitor prometheus-operator CRD
```
Nit: Rephrasing the title for clarity.
```suggestion
## Using the ServiceMonitor to access metrics with Prometheus
```

LiliC replied:
Not sure I agree with this, as ServiceMonitor is only a concept within prometheus-operator, I would include the name of it in the title.

hasbro17 replied:
Fair point. I think I've gotten used to associating it directly with Prometheus since the operator is always involved.

hasbro17 commented on doc/user/metrics/service-monitor.md:
```diff
@@ -0,0 +1,45 @@
+        
+        ...
+
+        // Populate bellow with Service(s) for which you want to create `ServiceMonitor` for.
```
Typo: *below
```suggestion
        // Populate below with the Service(s) for which you want to create ServiceMonitors.
```
Also no need to put the backticks for  \`ServiceMonitor \` since it's not being highlighted for markdown I think.

hasbro17 commented on doc/user/metrics/service-monitor.md:
```diff
@@ -0,0 +1,45 @@
+        // Change below value to name of the Namespace you want the `ServiceMonitor` to be created in.
+        ns := "default"
+        
+        // Pass the Service(s) to the helper function, which in turn returns the `ServiceMonitor` object.
```
It returns an array of ServiceMonitors right?
https://github.com/operator-framework/operator-sdk/blob/master/pkg/metrics/service-monitor.go#L29 
```suggestion
        // Pass the Service(s) to the helper function, which in turn returns the array of `ServiceMonitor` objects.
```

hasbro17 commented:
Perhaps this is outside the scope of this PR and we can do a follow up but I don't see any documentation on what the actual metrics being exposed are.
Should we list out what exactly are the controller related metrics being recorded by the controller-runtime? And perhaps give an example of querying the endpoint to see these metrics in an example?
https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg/internal/controller/metrics
https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/internal/controller/metrics/metrics.go#L50-L58
- Number of reconciliations
- Number of reconciliation errors per controller
- Reconciliation duration
- Resource metrics like CPU, memory usage, file descriptor usage
- Go runtime metrics like number of Go routines, GC duration

LiliC commented:
@hasbro17 Done, addressed the comments. I opened issue for the above suggestion to tackle in a follow up PR, when I am back from vacation. PTALA, thanks!
